syntax = "proto3";

package talent;

message PlotPhaseStatus {
    string stage = 1;
    float time = 2;
    float cpu_usage = 3;
}

message PlotP2BaseStatus {
    string stage = 1;
    PlotPhaseStatus scan = 2;
    PlotPhaseStatus sort = 3;
}

message PlotP1Status {
    string stage = 1;
    int32 table_1_now_size = 2;
    PlotPhaseStatus t1 = 3;
    PlotPhaseStatus t2 = 4;
    PlotPhaseStatus t3 = 5;
    PlotPhaseStatus t4 = 6;
    PlotPhaseStatus t5 = 7;
    PlotPhaseStatus t6 = 8;
    PlotPhaseStatus t7 = 9;
    int32 now = 10;
    float time = 11;
    float cpu_usage = 12;
}

message PlotP2Status {
    string stage = 1;
    PlotP2BaseStatus t7 = 2;
    PlotP2BaseStatus t6 = 3;
    PlotP2BaseStatus t5 = 4;
    PlotP2BaseStatus t4 = 5;
    PlotP2BaseStatus t3 = 6;
    PlotP2BaseStatus t2 = 7;
    PlotP2BaseStatus t1 = 8;
    int32 now = 9;
    float time = 10;
    float cpu_usage = 11;
}

message PlotP3Status {
    string stage = 1;
    PlotPhaseStatus t1_2 = 2;
    PlotPhaseStatus t2_3 = 3;
    PlotPhaseStatus t3_4 = 4;
    PlotPhaseStatus t4_5 = 5;
    PlotPhaseStatus t5_6 = 6;
    PlotPhaseStatus t6_7 = 7;
    string now = 8;
    float time = 9;
    float cpu_usage = 10;
}

message PlotDetails {
    string fpk = 1;
    string ppk = 2;
    string id = 3;
    int32 ksize = 4;
    string cache1 = 5;
    string cache2 = 6;
    int32 buffer = 7;
    int32 buckets = 8;
    int32 threads = 9;
    int32 stripe_size = 10;
    PlotP1Status phase_1_status = 11;
    PlotP2Status phase_2_status = 12;
    PlotP3Status phase_3_status = 13;
    PlotPhaseStatus phase_4_status = 14;
    PlotPhaseStatus total_time = 15;
    PlotPhaseStatus copy_time = 16;
    string dest_file_name = 17;
    string dest_path = 18;
    string dest_type = 19;
    int32 wrote = 20;
    int32 stage_now = 21;
    float progress = 22;
    string memo = 23;
    bool is_started = 24;
    bool is_finished = 25;
}

message PlotConfig {
    string fpk = 1;
    string ppk = 2;
    int32 ksize = 3;
    int32 threads = 4;
    int32 buffer = 5;
    string cache1 = 6;
    string cache2 = 7;
    message Dest {
        string type = 1;
        string path = 2;
    }
    Dest dest = 8;
}

message PlotTaskCreateRequest {
    string task_id = 1;
    string worker_id = 2;
    PlotConfig plot_config = 3;
}

message PlotTaskStatus {
    string worker_id = 1;
    string task_id = 2;
    bool existed = 3;
    int32 plot_pid = 4;
    int32 log_pid = 5;
    string status = 6;
    double received_time = 7;
    double pending_time = 8;
    double started_time = 9;
    double running_time = 10;
    double finished_time = 11;
    string remarks = 12;
    PlotDetails plot_details = 13;
}

message PlotTaskStatusAllResponse {
    repeated PlotTaskStatus tasks = 1;
}

message PlotTaskIdRequest {
    string task_id = 1;
}

message PlotTaskStopRequest {
    string task_id = 1;
    string reason = 2;
}

message GetPlotTaskResponse {
    // 所有未执行的任务列表
    repeated PlotConfig task_list = 1;
}

message PlotTaskStatusResponse {
    string type = 1;
    bool is_success = 2;
    string msg = 3;
}

message PlotTaskUpdateResponse {
    bool is_success = 1;
    string msg = 2;
}

message Empty {
    // 空参数必须用一个空消息
}

message GetPlotByCacheRequest {
    string status = 1;
    string cache1 = 2;
}

message PlotStatus {
    string status = 1;
}

service PlotManager {
    rpc plot_task_create(PlotTaskCreateRequest) returns (PlotTaskStatusResponse);
    rpc plot_task_status(PlotTaskIdRequest) returns (PlotTaskStatus);
    rpc plot_task_status_all(Empty) returns (PlotTaskStatusAllResponse);
    rpc plot_task_stop(PlotTaskStopRequest) returns (PlotTaskStatusResponse);
    rpc plot_task_update(PlotTaskStatus) returns (PlotTaskUpdateResponse);
    rpc get_plot_tasks(PlotStatus) returns (PlotTaskStatusAllResponse);
    rpc get_plot_by_cache(GetPlotByCacheRequest) returns (PlotTaskStatusAllResponse);
}




